packages/block-editor/src/store/reducer.js

```diff
@@ -2258,15 +2280,47 @@ function getDerivedBlockEditingModesForTree(
 	// so the default block editing mode is set to disabled.
 	const sectionRootClientId = state.settings?.[ sectionRootClientIdKey ];
 	const sectionClientIds = state.blocks.order.get( sectionRootClientId );
-	const syncedPatternClientIds = Object.keys(
-		state.blocks.controlledInnerBlocks
-	).filter(
-		( clientId ) =>
-			state.blocks.byClientId?.get( clientId )?.name === 'core/block'
+	const hasDisabledBlocks = Array.from( state.blockEditingModes ).some(
+		( [ , mode ] ) => mode === 'disabled'
 	);
+	const templatePartClientIds = [];
+	const syncedPatternClientIds = [];
+
+	Object.keys( state.blocks.controlledInnerBlocks ).forEach( ( clientId ) => {
+		const block = state.blocks.byClientId?.get( clientId );
+
+		if ( block?.name === 'core/template-part' ) {
+			templatePartClientIds.push( clientId );
+		}
+
+		if ( block?.name === 'core/block' ) {
+			syncedPatternClientIds.push( clientId );
+		}
+	} );
 
 	traverseBlockTree( state, treeClientId, ( block ) => {
 		const { clientId, name: blockName } = block;
+
+		// If the block already has an explicit block editing mode set,
+		// don't override it.
+		if ( state.blockEditingModes.has( clientId ) ) {
+			return;
+		}
+
+		// An explicitly disabled block editing mode cascades to all its children.
+		// Check ancestors of the current block to see if the nearest one with a block
+		// editing mode set is disabled.
+		if ( hasDisabledBlocks ) {
+			const ancestorBlockEditingMode = getAncestorBlockEditingMode(
+				state,
+				clientId
+			);
+			if ( ancestorBlockEditingMode === 'disabled' ) {
+				derivedBlockEditingModes.set( clientId, 'disabled' );
+				return;
+			}
+		}
```

Comment:
This is the part I've refactored from the `getBlockEditingModes` selector.

After this change, if a parent (ancestor) block has a `blockEditingMode` of 'disabled', the children will have a `disabled` derivedBlockEditingMode to reflect the way the mode is inherited by children.

The selector already has code to check `derivedBlockEditingModes` and return the value from it.